Add directive to import layers from a WMTS server

Jenselme opened the pull request:
I have a working solution that looks good to me. However, I have an error I don't know how to correctly fix when I run `make check`:

```
node buildtools/build.js .build/examples/interactionbtngroup.json .build/examples/interactionbtngroup.js
info ol Parsing dependencies
ERR! Unsatisfied dependency "ol.source.WMTS.optionsFromCapabilities" in script: /home/jenselme/Work/forks/ngeo/src/modules/import/wmtsgetcap.js 
Makefile:346: recipe for target '.build/examples/interactionbtngroup.js' failed
make: *** [.build/examples/interactionbtngroup.js] Error 1
```

I require `ol.source.WMTS.optionsFromCapabilities` in [wmtsgetcap.js](https://github.com/Jenselme/ngeo/blob/wmts-import/src/modules/import/wmtsgetcap.js#L5) because I use it [here](https://github.com/Jenselme/ngeo/blob/wmts-import/src/modules/import/wmtsgetcap.js#L42). How can I require it without causing any issues during the build?

Close #2537

gberaudo commented:
Have you tried requiring `ol.source.WMTS` instead?

Jenselme commented:
@gberaudo I did not. Just tried it and it worked. Thanks for the solution. This looks ready for a review.

gberaudo commented on src/modules/import/wmtsgetcap.js:
```diff
@@ -0,0 +1,145 @@
+
+      if (!layer['isInvalid']) {
+        const requestEncoding = getCap['OperationsMetadata']['GetTile']['DCP']
+          .HTTP
```
You need quotes everywhere.

gberaudo commented on src/modules/import/wmtsgetcap.js:
```diff
@@ -0,0 +1,145 @@
+          .Value[0];
+        const layerOptions = {
+          layer: layer['Identifier'],
+          requestEnconding: requestEncoding
```
You need quotes here.

gberaudo commented on src/modules/import/wmtsgetcap.js:
```diff
@@ -0,0 +1,145 @@
+          .DCP
+          .HTTP
+          .Get[0]
+          .href;
```
You need quotes everywhere.

gberaudo commented on src/modules/import/wmtsgetcap.js:
```diff
@@ -0,0 +1,145 @@
+    /**
+     * @param {angular.JQLite} element Element.
+     * @param {angular.Attributes} attrs Attributes.
+     * @return {boolean} Template URL.
```
It returns a string.

Jenselme replied:
True. I made a copy paste from [wmsgetcap](https://github.com/camptocamp/ngeo/blob/master/src/modules/import/wmsgetcap.js#L201) where it says boolean. Should I correct the type there as well?

gberaudo replied:
Yes, sure

gberaudo commented on src/modules/import/wmtsgetcapitem.js:
```diff
@@ -0,0 +1,86 @@
+          elt.append(clone);
+        });
+      };
+    }
```
Not sure what this compile fuction do. Would it be possible to use an Angular component instead of an Angular directive?

Jenselme replied:
I don't know either. I took my inspiration from [wmsgetcapitem](https://github.com/camptocamp/ngeo/blob/master/src/modules/import/wmsgetcapitem.js#L126) I have never seen this before.

oterral replied:
It's used to construct a recursive tree: http://stackoverflow.com/questions/14430655/recursion-in-angular-directives/18609594

gberaudo commented on src/modules/import/wmtsgetcapitem.js:
```diff
@@ -0,0 +1,86 @@
+    /**
+     * @param {angular.JQLite} element Element.
+     * @param {angular.Attributes} attrs Attributes.
+     * @return {boolean} Template URL.
```
It is a string.

Jenselme replied:
True. I made a copy paste from [wmsgetcapitem](https://github.com/camptocamp/ngeo/blob/master/src/modules/import/wmsgetcapitem.js#L157) where it says boolean. Should I correct the type there as well?

fredj commented on src/modules/import/wmtsgetcap.js:
```diff
@@ -0,0 +1,145 @@
+
+/**
+ * @constructor
+ * @param {Window} $window The window.
```
no need to use `$window`, simply use `console.error(...)` and `alert(...)`

Jenselme replied:
OK. It is the same in wmsgetcap. Should I correct this there as well?

Jenselme commented:
PR updated to take your comments into account. I can squash the commit if you like.

gberaudo commented:
Thanks, there is one more thing to do before merging, you should update the contribs/gmf/apps/desktop_alt/index.html application to make use of the new feature.

Then I will be able to see it live and merge.

Jenselme commented:
@gberaudo just added an example in desktop_alt. Let me know what you think.

gberaudo commented:
@Jenselme, works great in debug mode but in hosted mode (http://localhost:3000/.build/examples-hosted/contribs/gmf/apps/desktop_alt/index.html) the layer does not appear.

Jenselme commented:
@gberaudo that's possible. I rely on a WMTS server by swisstopo. If your domain is localhost, you can use the tiles (debug mode) but if it is not, you need to register your domain. Otherwise, the requests for tiles will fail with a 403 status code.

Maybe I should use another WMTS server for the example. Do you know one I could use?

fredj commented:
What about https://ows.asitvd.ch/wmts1.0.0/WMTSCapabilities.xml ?

Jenselme commented:
@fredj This should work. PR updated to use this WMTS server.

rbovard commented on contribs/gmf/apps/desktop_alt/js/importhelper.js:
```diff
@@ -28,6 +28,7 @@ app.GmfImportHelper = function(map, $scope, gettext, ngeoFile, $q) {
 
   this.urls = [
     {'name': 'geoadmin', 'url': 'https://wms.geo.admin.ch/'},
+    {'name': 'geoadmin WMTS', 'url': 'https://ows.asitvd.ch/wmts1.0.0/WMTSCapabilities.xml'},
```
'geoadmin WMTS' > 'ASIT VD WMTS'

Jenselme replied:
@rbovard oops! Nice catch, corrected.

gberaudo commented:
Thanks for the update @Jenselme. I tested again both in debug and hosted modes, there are no more CORS errors but unfortunately the layer still does not show up in hosted mode.
It really looks like a minification error.

Jenselme commented:
@gberaudo It was indeed a minification error. I used `getCapLayer.capabilitiesUrl` instead of `getCapLayer['capabilitiesUrl']`. I just corrected that and updated the PR.

Jenselme commented:
@gberaudo it looks like one of the test is timing out on Travis: https://travis-ci.org/camptocamp/ngeo/jobs/231096275

fredj commented:
I've restarted the build

Jenselme commented:
@fredj thanks but it timed out again.

fredj commented:
let's try again ...

Jenselme commented:
@fredj all green now. Thanks.

gberaudo commented:
Thanks for your contribution @Jenselme.

Jenselme commented:
@gberaudo @fredj thanks for the review.
